docs/_docs/rql.md

```diff
@@ -0,0 +1,470 @@
+
+```
+PE <PredicateType> :=
+  [BinaryOp, PE <PredicateType>, PE <PredicateType>] |
+  <PredicateType>
+
+BinaryOp := "AND" | "OR"
+```
+
+where `<PredicateType>` has its own AST grammar. Informally, `PE <PredicateType>` translates to `a predicate expression of <PredicateType>s`. For example, `PE Primary` translates to `a predicate expression of primaries`.
+
+Similarly, let `NPE <PredicateType>` denote the following grammar (where `NPE` => `NegatablePredicateExpression`):
+
+```
+NPE <PredicateType> :=
+  [UnaryOp,  NPE <PredicateType>]                       |
+  [BinaryOp, NPE <PredicateType>, NPE <PredicateType>]  |
+  <PredicateType>
+
+UnaryOp := ‘NOT’
+```
+
+Then the RQL AST can be expressed as
+
+```
+Query := PE Primary
+
+Primary :=
+  Action  |
+  Boolean |
+  Name    |
+  CName   |
+  Path    |
+  Kind    |
+  Atime   |
+  Crtime  |
+  Ctime   |
+  Mtime   |
+  Size    |
+  Meta
+
+Action  := [“action”, NPE ActionPredicate]
+Boolean := BooleanPredicate
+Name    := [“name”,   NPE StringPredicate]
+CName   := [“cname”,  NPE StringPredicate]
+Path    := [“path”,   NPE StringPredicate]
+Kind    := [“kind”,   NPE StringPredicate]
+Atime   := [“atime”,  NPE TimePredicate]
+Crtime  := [“crtime”, NPE TimePredicate]
+Ctime   := [“ctime”,  NPE TimePredicate]
+Mtime   := [“mtime”,  NPE TimePredicate]
+Size    := SizePredicate
+Meta    := [“meta”,   PE ObjectPredicate]
+
+ActionPredicate := 
+  "list"   |
+  "read"   |
+  "write"  |
+  "stream" |
+  "exec"   |
+  "delete"
+
+ValuePredicate :=
+  ObjectPredicate                    |
+  ArrayPredicate                     |
+  NullPredicate                      |
+  BooleanPredicate                   |
+  [“number”, NPE NumericPredicate]   |
+  [“time”,   NPE TimePredicate]      |
+  [“string”, NPE StringPredicate]
+
+ObjectPredicate        := [“object”, SizePredicate | ObjectElementPredicate]
+ObjectElementPredicate := [ObjectElementSelector, NPE ValuePredicate]
+ObjectElementSelector  :=  [“key”, [“=”, <string>]]
+
+ArrayPredicate        := [“array”, SizePredicate | ArrayElementPredicate]
+ArrayElementPredicate := [ArrayElementSelector, NPE ValuePredicate]]
+ArrayElementSelector  := “some” | “all”  | <array_index>
+
+SizePredicate := [“size”, NPE NumericPredicate (n >= 0)]
+
+NullPredicate := null
+
+BooleanPredicate := true | false
+
+NumericPredicate  := [ComparisonOp, Number]
+ComparisonOp      := “<” | “>” | “<=” | “>=” | “=” | “!=”
+
+StringPredicate :=
+  [“glob”,   <glob>]   |
+  [“regex”,  <regex>]  |
+  [“=”,      <string>]
+
+TimePredicate := [ComparisonOp, TimeValue]
+```
+
+This grammar tells us several things:
+
+* An RQL query is a predicate expression of primaries. A primary is a predicate on a specific field of an entry. Examples of primaries include `name`, `cname`, and `ctime`. See the [primaries](#primaries) section for all of the available primaries.
+
+* Primaries take predicates, specifically predicate expressions.
```

Comment:
I think I'd put a few small examples here before getting into specific primaries.

Reply:
You want to take what they just read and map it into some concrete examples that help to show the simple patterns it represents.